### ISSUE TITLE
optional outputs: expire/succeeded not orthogonal?

### ISSUE THREAD

**Issue by oliver-sanders** [cylc/cylc-flow#5361]
This infinite workflow:

```ini
[scheduling]         
    initial cycle point = previous(T00)    
    runahead limit = P1    
    [[special tasks]]    
        clock-expire = foo(PT0S)    
    [[graph]]        
        R1 = """    
            slow => foo    
        """                                                                                                                        
        PT1H = """    
            foo[-PT1H] => foo    
            foo => bar    
        """          
                     
[runtime]        
    [[slow]]     
        script = sleep 3    
    [[foo]]      
    [[bar]]  
```

Will shut down on the first cycle with no error:

```log
INFO - Extracting job.sh to /net/home/h06/osanders/cylc-run/clock-expire/run55/.service/etc/job.sh
INFO - Workflow: clock-expire/run55
INFO - Scheduler: url=tcp://vld601.cmpd1.metoffice.gov.uk:43144 pid=48963
INFO - Workflow publisher: url=tcp://vld601.cmpd1.metoffice.gov.uk:43110
INFO - Run: (re)start number=1, log rollover=1
INFO - Cylc version: 8.1.1
INFO - Run mode: live
INFO - Initial point: 20230210T0000Z
INFO - Final point: None
INFO - Cold start from 20230210T0000Z
INFO - New flow: 1 (original flow from 20230210T0000Z) 2023-02-10 10:55:26
INFO - [20230210T0000Z/slow waiting(runahead) job:00 flows:1] => waiting
INFO - [20230210T0000Z/slow waiting job:00 flows:1] => waiting(queued)
INFO - [20230210T0000Z/slow waiting(queued) job:00 flows:1] => waiting
INFO - [20230210T0000Z/slow waiting job:01 flows:1] => preparing
INFO - [20230210T0000Z/slow preparing job:01 flows:1] submitted to localhost:background[49000]
INFO - [20230210T0000Z/slow preparing job:01 flows:1] => submitted
INFO - [20230210T0000Z/slow submitted job:01 flows:1] health: submission timeout=P1D, polling intervals=PT1H,...
INFO - [20230210T0000Z/slow submitted job:01 flows:1] => running
INFO - [20230210T0000Z/slow running job:01 flows:1] health: execution timeout=None, polling intervals=PT1H,...
INFO - [20230210T0000Z/slow running job:01 flows:1] => succeeded
INFO - [20230210T0000Z/foo waiting(runahead) job:00 flows:1] => waiting
INFO - [20230210T0000Z/foo waiting job:00 flows:1] => waiting(queued)
WARNING - [20230210T0000Z/foo waiting(queued) job:00 flows:1] Task expired (skipping job).
INFO - [20230210T0000Z/foo waiting(queued) job:00 flows:1] => expired(queued)
INFO - Workflow shutting down - AUTOMATIC
INFO - DONE
```

Presumably because expiry skips the compulsory output check preventing the non-expired future tasks from being spawned.

**Pull requests welcome!**
This is an Open Source project - please consider contributing a bug fix
yourself (please read `CONTRIBUTING.md` before starting any work though).

**Comment by hjoliver** [comment]
I'm not sure why you think this is a bug.

```
foo[-PT1H] => foo  
```

The graph says, if `1/foo` expired, then `2/foo` will not be triggered, which brings this branch of the graph to an end. And in this case, there are no other branches so the scheduler shuts down.

Perhaps you really want this?:
```
foo[-PT1H] | foo[-PTH]:expired => foo 
```

(Which doesn't work either, but that's because the `:expired` trigger is broken!)

**Comment by oliver-sanders** [comment]
> I'm not sure why you think this is a bug.

Two reasons:

1. The "succeeded" compulsory output was *not* produced, but the scheduler did not detect or report this.
2. The scheduler should have stalled because of this unsatisfied dependency `foo[-PT1H] => foo`.

**Comment by oliver-sanders** [comment]
> Perhaps you really want this?:
>
> foo[-PT1H] | foo[-PTH]:expired => foo 

Shouldn't that be:

```
foo[-PT1H]? | foo[-PTH]:expired? => foo 
```

**Comment by hjoliver** [comment]
> The "succeeded" compulsory output was not produced, but the scheduler did not detect or report this.

Ah, right. Yes that is a bug.  It's really another symptom of the over-enthusiastic removal of expired tasks: #5364  ... I'll add a note to that issue.

**Comment by oliver-sanders** [comment]
Superseded by https://github.com/cylc/cylc-flow/issues/5640
